File: forge/src/main/java/com/eliasnvx/krylix/forge/MobKillStatsData.java
package com.eliasnvx.krylix.forge;

import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.server.MinecraftServer;
import net.minecraft.world.level.saveddata.SavedData;

import java.util.HashMap;
import java.util.Map;

public class MobKillStatsData extends SavedData {
    public final Map<String, Integer> kills = new HashMap<>();
    private static final String ID = "krylix_mob_stats";

    public Map<String, Integer> increment(String entityId) {
        kills.put(entityId, kills.getOrDefault(entityId, 0) + 1);
        setDirty();
        return kills;
    }

    @Override
    public CompoundTag save(CompoundTag tag, HolderLookup.Provider registries) {
        CompoundTag killsTag = new CompoundTag();
        for (Map.Entry<String, Integer> entry : kills.entrySet()) {
            killsTag.putInt(entry.getKey(), entry.getValue());
        }
        tag.put("kills", killsTag);
        return tag;
    }

    private static MobKillStatsData load(CompoundTag tag, HolderLookup.Provider registries) {
        MobKillStatsData data = new MobKillStatsData();
        CompoundTag killsTag = tag.getCompound("kills");
        for (String key : killsTag.getAllKeys()) {
            data.kills.put(key, killsTag.getInt(key));
        }
        return data;
    }

    public static MobKillStatsData get(MinecraftServer server) {
        SavedData.Factory<MobKillStatsData> factory = new SavedData.Factory<>(MobKillStatsData::new, MobKillStatsData::load, null);
        return server.overworld().getDataStorage().computeIfAbsent(factory, ID);
    }
}

File: forge/src/main/kotlin/com/eliasnvx/krylix/forge/MobKillStatsData.kt
package com.eliasnvx.krylix.forge

import net.minecraft.nbt.CompoundTag
import net.minecraft.server.MinecraftServer
import net.minecraft.world.level.saveddata.SavedData

/**
 * Хранит количество убийств враждебных мобов за весь мир (per-world, не per-dimension —
 * всегда читается/пишется через оверворлд текущего сервера).
 */
class MobKillStatsData : SavedData() {
    val kills: MutableMap<String, Int> = mutableMapOf()

    fun increment(entityId: String): Map<String, Int> {
        kills[entityId] = (kills[entityId] ?: 0) + 1
        setDirty()
        return kills
    }

    override fun save(tag: CompoundTag): CompoundTag {
        val killsTag = CompoundTag()
        kills.forEach { (id, count) -> killsTag.putInt(id, count) }
        tag.put("kills", killsTag)
        return tag
    }

    companion object {
        private const val ID = "krylix_mob_stats"

        private fun load(tag: CompoundTag): MobKillStatsData {
            val data = MobKillStatsData()
            val killsTag = tag.getCompound("kills")
            for (key in killsTag.allKeys) {
                data.kills[key] = killsTag.getInt(key)
            }
            return data
        }

        fun get(server: MinecraftServer): MobKillStatsData =
            server.overworld().dataStorage.computeIfAbsent(::load, ::MobKillStatsData, ID)
    }
}
